pkg/shares/share_builder.go

```diff
@@ -0,0 +1,205 @@
+
+	// We need to finish this share and start a new one
+	// so we return the leftover to be written into a new share
+	return rawData[pendingLeft:]
+}
+
+func (b *Builder) Build() (*Share, error) {
+	return NewShare(b.rawShareData)
+}
+
+// IsEmptyShare returns true if no data has been written to the share
+func (b *Builder) IsEmptyShare() bool {
+	if b.isFirstShare {
+		return len(b.rawShareData) == appconsts.NamespaceSize+appconsts.ShareInfoBytes+appconsts.SequenceLenBytes+appconsts.CompactShareReservedBytes
+	}
+	return len(b.rawShareData) == appconsts.NamespaceSize+appconsts.ShareInfoBytes+appconsts.CompactShareReservedBytes
+}
+
+func (b *Builder) ZeroPadIfNecessary() (bytesOfPadding int) {
+	b.rawShareData, bytesOfPadding = zeroPadIfNecessary(b.rawShareData, appconsts.ShareSize)
+	return
+}
+
+// isEmptyReservedBytes returns true if the reserved bytes are empty.
+func (b *Builder) isEmptyReservedBytes() bool {
+	indexOfReservedBytes := b.indexOfReservedBytes()
+	reservedBytes, err := ParseReservedBytes(b.rawShareData[indexOfReservedBytes : indexOfReservedBytes+appconsts.CompactShareReservedBytes])
+	if err != nil {
+		panic(err)
+	}
+	return reservedBytes == 0
+}
+
+// indexOfReservedBytes returns the index of the reserved bytes in the share.
+func (b *Builder) indexOfReservedBytes() int {
+	if b.isFirstShare {
+		// if the share is the first share, the reserved bytes follow the namespace, info byte, and sequence length
+		return appconsts.NamespaceSize + appconsts.ShareInfoBytes + appconsts.SequenceLenBytes
+	}
+	// if the share is not the first share, the reserved bytes follow the namespace and info byte
+	return appconsts.NamespaceSize + appconsts.ShareInfoBytes
+}
+
+// indexOfInfoBytes returns the index of the InfoBytes.
+func (b *Builder) indexOfInfoBytes() int {
+	// the info byte is immediately after the namespace
+	return appconsts.NamespaceSize
+}
+
+// MaybeWriteReservedBytes will be a no-op if the reserved bytes
+// have already been populated. If the reserved bytes are empty, it will write
+// the location of the next unit of data to the reserved bytes.
+func (b *Builder) MaybeWriteReservedBytes() error {
+	if !b.isCompactShare {
+		return errors.New("this is not a compact share")
+	}
+
+	if !b.isEmptyReservedBytes() {
+		return nil
+	}
+
+	byteIndexOfNextUnit := len(b.rawShareData)
+	reservedBytes, err := NewReservedBytes(uint32(byteIndexOfNextUnit))
+	if err != nil {
+		return err
+	}
+
+	indexOfReservedBytes := b.indexOfReservedBytes()
+	// overwrite the reserved bytes of the pending share
+	for i := 0; i < appconsts.CompactShareReservedBytes; i++ {
+		b.rawShareData[indexOfReservedBytes+i] = reservedBytes[i]
+	}
+	return nil
+}
+
+// writeSequenceLen writes the sequence length to the first share.
+func (b *Builder) WriteSequenceLen(sequenceLen uint32) error {
+	if !b.isFirstShare {
+		return errors.New("not the first share")
+	}
+	if b == nil {
+		return errors.New("the builder object is not initialized (is nil)")
+	}
```

Comment:
[question] when will this error be thrown? If the builder object isn't initialized, the error above will be thrown.

```go
func TestWriteSequenceLen(t *testing.T) {
	b := Builder{}
	got := b.WriteSequenceLen(10)
	fmt.Printf("got: %v", got) // got: not the first share
	assert.Error(t, got)
}
```